File: services/server/src/safe_socket/safe_socket.py
import socket
import struct 

def recv_all(socket: socket.socket, size: int) -> bytes:
    data = bytearray()

    while len(data) < size:
        try:
            chunk = socket.recv(max(size - len(data), 2))
        except InterruptedError:
            continue
        except OSError as e:
            raise ConnectionError(f"recv failed: {e}") from e

        if not chunk:
            if len(data) == 0:
                raise EOFError("socket closed by peer")
            raise ConnectionError(
                f"socket connection closed: expected={size} received={len(data)}"
            )

        data.extend(chunk)

    return bytes(data)

def send_all(socket: socket.socket, bytes_data: bytes) -> int:
    total_sent = 0

    while total_sent < len(bytes_data):
        try:
            sent = socket.send(bytes_data[total_sent:])
        except InterruptedError:
            continue
        except OSError as e:
            raise ConnectionError(f"send failed: {e}") from e

        total_sent += sent

    return total_sent

def recv_message(socket: socket.socket):
    message_len_bytes = recv_all(socket, 4)

    if len(message_len_bytes) != 4:
        raise ConnectionError("Incomplete message length")

    message_len = struct.unpack("!I", message_len_bytes)[0]

    message = recv_all(socket, message_len)

    if len(message) != message_len:
        raise ConnectionError("Incomplete message")

    return message

def send_message(socket: socket.socket, message: bytes):
    packet = struct.pack("!I", len(message)) + message
    send_all(socket, packet)

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"encoding/binary"
	"fmt"
	"io"
)

const headerSize = 4

func SendAll(socket io.Writer, bytes []byte) error {
	totalSent := 0

	for totalSent < len(bytes) {
		sent, err := socket.Write(bytes[totalSent:])

		if sent > 0 {
			totalSent += sent
		}

		if err != nil {
			return err
		}
	}

	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalReceived := 0

	for totalReceived < size {
		received, err := socket.Read(buff[totalReceived:])

		if received > 0 {
			totalReceived += received
		}

		if err != nil {
			if err == io.EOF {
				if totalReceived == size {
					break
				}
				return nil, fmt.Errorf("socket connection closed: expected=%d received=%d", size, totalReceived)
			}
			return nil, err
		}

		if received == 0 {
			return nil, fmt.Errorf("socket connection closed: expected=%d received=%d", size, totalReceived)
		}
	}

	return buff, nil
}

func SendMessage(socket io.Writer, message []byte) error {
	header := [headerSize]byte{}
	binary.BigEndian.PutUint32(header[:], uint32(len(message)))

	if err := SendAll(socket, header[:]); err != nil {
		return err
	}
	return SendAll(socket, message)
}

func RecvMessage(socket io.Reader) ([]byte, error) {
	messageLengthBytes, err := RecvAll(socket, headerSize)
	if err != nil {
		return nil, err
	}

	if len(messageLengthBytes) != headerSize {
		return nil, fmt.Errorf("incomplete message length")
	}

	messageLength := binary.BigEndian.Uint32(messageLengthBytes)

	message, err := RecvAll(socket, int(messageLength))
	if err != nil {
		return nil, err
	}

	if len(message) != int(messageLength) {
		return nil, fmt.Errorf("incomplete message")
	}

	return message, nil
}
